README
This is a "plugin" for the Video Disk Recorder (VDR).

Written by:                           Hardy Flor <HFlor@gmx.de>

Project's homepage (vdr):             http://www.tvdr.de/
Project's homepage (playlist):        http://projects.vdr-developer.org/projects/plg-playlist

See the file COPYING for license information.

playlist-plugin version 0.0.2rc3

hier kommt noch was ...

TODO:

- Komandozeilenoptionen und Konfigurationsdatei fehlen noch
- Markierung in Aufnameauswahl
- Verschiebung in Liste genauer Prüfen (z.Z. kann man Verzeichnisse verschieben --> ist aber nicht gut!)
- Aufnahmenanzeige nicht im richtigen Verzeichnis bei vorheriger Auswahl eines Verzeichnisses mit Unterverzeichnissen
- Aufnahmenanzeige enthält doppelte Einträge
- auf roter taste standartwert anzeigen
- Option für Wiedergabe im Detail-Ansicht 
x neue Playlist ohne inhalt auch namen-dialog schließen
- Wenn nur 1 Verzeichniss in PL -> Name für PL übernehmen



HISTORY
VDR Plugin 'playlist' Revision History
--------------------------------------

2004-09-23: Version 0.0.1

- Initial revision.

2004-10-08: Version 0.0.2

* Bug fixes

2012-07-01 Martin Dummer <martin.dummer@gmx.net>

- edit the README file: adjust mail addresses and web URLs
- converted README von ISO-8859-1 to UTF-8
- merge patches from debian source-packages from http://e-Tobi.net
    Adds -fPIC to Makefile to fix potential FTBFS
    Support for the APIVERSION define as introduced in VDR-1.3.47 by Thomas Günther <tom@toms-cafe.de>
    Changes for VDR >= 1.3.25 and VDR >= 1.3.36 by viking at vdrportal.de
    Changes for VDR >= 1.5.7 by Thomas Günther <tom@toms-cafe.de>
    Changes for VDR >= 1.7.3 by Thomas Günther <tom@toms-cafe.de>

2012-07-02 Martin Dummer <martin.dummer@gmx.net>

- convert i18n handling to gettext



playlist.c
/*
 * playlist: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id: playlist.c 0.2 2004/10/08 02:34:00 hflor Exp $
 */

#include "playlist.h"
#include "menuplaylists.h"
#include "menucontrol.h"
#include "menusetup.h"
#include "vdrtools.h"
#include <getopt.h>
#include <ctype.h>
#include <vdr/recording.h>

static const char *VERSION           = "0.0.2";
static const char *DESCRIPTION       = "playlist for recordings";

/* TODO
option playonlynewrecords
*/
// Global variables that control the overall behaviour:

// var pos2
tParamFile  playlistconfigfile       = { false, false, false, "playlist.conf", "playlist.conf" };
#ifdef HAVE_ICONPATCH
tParamChar  charentry                = { false, false, false, 236 };
tParamChar  charfolder               = { false, false, false, 237 };
tParamChar  charrecord               = { false, false, false, 249 };
#else
tParamChar  charentry                = { false, false, false, 'e' };
tParamChar  charfolder               = { false, false, false, 'f' };
tParamChar  charrecord               = { false, false, false, 'r' };
#endif
tParamInt   commandline_preference   = { false, false, false, true };
tParamFile  lastplaylist             = { false, false, false, "" };
tParamFile  mainmenu_name            = { false, false, false, "Playlist" };
tParamInt   mainmenu_visible         = { false, false, false, true };
tParamInt   confirmdelplentry        = { false, false, false, true };
tParamInt   okstartplay              = { false, false, false, true };
tParamInt   deleterecords            = { false, false, false, true };
tParamFile  pathplaylists            = { false, false, false, "" };
tParamInt   startoption              = { false, false, false, 0 };
tParamInt   storeplaylist            = { false, false, false, 0 };
tParamInt   timeoutreadrecords       = { false, false, false, 60 };
tParamInt   verbose                  = { false, false, false, false, true };
tParamInt   PL_options[Option_max]   ={{ false, false, false, true },                  // confirmdeleterecord
                                       { false, false, false, false },                 // confirmstartnewrecord
                                       { false, false, false, true },                  // deletefromlist
                                       { false, false, false, false },                 // deleterecord
                                       { false, false, false, true },                  // jumpmark
                                       { false, false, false, PlayOptions_firstnew },  // playoption1
                                       { false, false, false, PlayOptions_selectpos }, // playoption2
                                       { false, false, false, false },                 // playonlynew
                                       { false, false, false, true },                  // searchnewrecord
                                       { false, false, false, true }};                 // searchrecordsub

char        plugin_name[MaxFileName]  = "Playlist";
char        DisplayStatus[MAXOSDTEXTWIDTH] = "";
const char  *NoYesDefault[NoYesDefault_max];
const char  *StartOptions[Start_max];
const char  *PlayOptions[PlayOptions_max];
char        *FileNameCharsAllowed = NULL;
char        *AllCharsAllowed = NULL;
char        *LastSelectedRecord = NULL;
bool        ControlMenuIsOpen = false;
#if VDRVERSNUM < 10311
cRecordings Recordings;
time_t      LoadTime_Recordings;
#endif

void DisplaySetings(void)
{
  #define WriteSource(T)   plugin_name, T.r ? 'r' : ' ', T.c ? 'c' : ' '
  #define IsDefault(T)     WriteSource(T), T.u == T.d ? '*' : ' '
  #define IsDefaultChar(T) WriteSource(T), strcmp(T.u, T.d) ? ' ' : '*'
  #define BoolValue(T)     IsDefault(T), T.u ? "yes" : "no"

  // var pos3
  ExpandEnvironment(&pathplaylists);
  if (verbose.u)
  {
    isyslog("%s: commandline_preference = [ %c%c%c ] %s", BoolValue(commandline_preference));
    isyslog("%s: mainmenu_visible       = [ %c%c%c ] %s", BoolValue(mainmenu_visible));
    if (mainmenu_visible.u)
      isyslog("%s: mainmenu_name          = [ %c%c%c ] %s", IsDefaultChar(mainmenu_name), mainmenu_name.u);
  }

#ifdef PL_Debug1
  if (!verbose.u)
  {
    dsyslog("%s: commandline_preference = [ %c%c%c ] %s", BoolValue(commandline_preference));
    dsyslog("%s: mainmenu_visible       = [ %c%c%c ] %s", BoolValue(mainmenu_visible));
    if (mainmenu_visible.u)
      dsyslog("%s: mainmenu_name          = [ %c%c%c ] %s", IsDefaultChar(mainmenu_name), mainmenu_name.u);
  }
#endif

  #undef WriteSource
  #undef IsDefault
  #undef IsDefaultChar
  #undef BoolValue
}

void ExpandEnvironment(tParamFile *FileStruc)
{
#ifdef PL_Debug2
  dsyslog("%s: ExpandEnvironment text=%s", plugin_name, FileStruc->u);
#endif
  char *s;
  char *p;
  strn0cpy(FileStruc->e, FileStruc->u, sizeof(FileStruc->e));
  while ((s = strstr(FileStruc->e, "$(")) || (s = strstr(FileStruc->e, "${")))
  {
    char c = *(s + 1) == '(' ? ')' : '}';
    p = strchr(s, c); // search closing ) or }
    if (p)
    {
      *p++ = 0;
      *s = 0;
      s += 2;
      char *e = getenv(s);
      if (e)
      {
        char *buffer = NULL;
        asprintf(&buffer, "%s%s%s", FileStruc->e, e, p);
        strn0cpy(FileStruc->e, buffer, sizeof(FileStruc->e));
        FREENULL(buffer);
      } else
      {
        esyslog("%s: environmentvariable '%s' not found path=%s", plugin_name, s, FileStruc->u);
        FileStruc->e[0] = 0;
      }
    } else
    {
      esyslog("%s: missing ')' after '$(' path=%s", plugin_name, FileStruc->u);
      FileStruc->e[0] = 0;
    }
  }
  while ((p = strstr(FileStruc->e, "//")))
    strcpy(p, p + 1);
#ifdef PL_Debug2
  dsyslog("%s: ExpandEnvironment return=%s", plugin_name, FileStruc->e);
#endif
}

// --- cPluginPlaylist ----------------------------------------------------------

cPluginPlaylist *PluginPlaylist;

cPluginPlaylist::cPluginPlaylist(void)
{
#ifdef PL_Debug1
  dsyslog("%s: cPluginPlaylist::cPluginPlaylist", plugin_name);
#endif
  // Initialize any member variables here.
  // DON'T DO ANYTHING ELSE THAT MAY HAVE SIDE EFFECTS, REQUIRE GLOBAL
  // VDR OBJECTS TO EXIST OR PRODUCE ANY OUTPUT!
#if VDRVERSNUM >= 10507
  OSDLanguage = NULL;
#else
  OSDLanguage = -1;
#endif
  PluginPlaylist = this;
}

cPluginPlaylist::~cPluginPlaylist()
{
#ifdef PL_Debug1
  dsyslog("%s: cPluginPlaylist::~cPluginPlaylist", plugin_name);
#endif
  // Clean up after yourself!
  FREENULL(FileNameCharsAllowed);
  FREENULL(AllCharsAllowed);
  FREENULL(LastSelectedRecord);
  PluginPlaylist = NULL;
}

void cPluginPlaylist::TestAndSetOSDLanguage(void)
{
#ifdef PL_Debug2
  dsyslog("%s: cPluginPlaylist::TestAndSetOSDLanguage OSDLanguage=%d", plugin_name, Setup.OSDLanguage);
#endif
  if (OSDLanguage != Setup.OSDLanguage)
  {
    OSDLanguage = Setup.OSDLanguage;

    StartOptions[Start_DisplayPL] = tr("Setup$Display PL");
    StartOptions[Start_DisplayLast] = tr("Setup$Display last PL");
    StartOptions[Start_PlayLast] = tr("Setup$Play last PL");
    StartOptions[Start_NewEmpty] = tr("Setup$new empty PL");
    NoYesDefault[NoYesDefault_no] = tr("no");
    NoYesDefault[NoYesDefault_yes] = tr("yes");
    NoYesDefault[NoYesDefault_defaultPlugin] = tr("OptionPL$[from Setup]");
    NoYesDefault[NoYesDefault_defaultPlaylist] = tr("OptionPL$[from Playlist]");
    NoYesDefault[NoYesDefault_defaultRecordDir] = tr("OptionPL$[from Folder]");
    PlayOptions[PlayOptions_firstpos] = tr("OptionPL$first pos");
    PlayOptions[PlayOptions_firstnew] = tr("OptionPL$first new");
    PlayOptions[PlayOptions_lastplay] = tr("OptionPL$last played");
    PlayOptions[PlayOptions_selectpos] = tr("OptionPL$seletced pos");
    PlayOptions[PlayOptions_question] = tr("OptionPL$question");
    PlayOptions[PlayOptions_defaultPlugin] = tr("OptionPL$[from Setup]");
    FREENULL(FileNameCharsAllowed);
    asprintf(&FileNameCharsAllowed, "%s/$(){}!%%@", tr(FileNameChars));
    if (!AllCharsAllowed)
    {
      AllCharsAllowed = MALLOC(char, 193);
      char *next = AllCharsAllowed;
      for (int i = 32; i <= 127; i++)
        *next++ = i;
      for (int i = 160; i <= 255; i++)
        *next++ = i;
      *next = 0;
    }
  }
}

bool cPluginPlaylist::ProcessArg(int argc, char *argv[])
{
  int c;
  static struct option long_options[] = {
    { "delete_begin",              no_argument,       NULL, 'b' },
    { "nodelete_begin",            no_argument,       NULL, 'B' },
    { "min_entrys",                required_argument, NULL, 'd' },
    { "delete_end",                no_argument,       NULL, 'e' },
    { "nodelete_end",              no_argument,       NULL, 'E' },
    { "holdtime_history",          required_argument, NULL, 'h' },
    { "visible_in_mainmenu",       no_argument,       NULL, 'm' },
    { "hide_in_mainmenu",          no_argument,       NULL, 'M' },
    { "mainmenu_name",             required_argument, NULL, 'n' },
    { "holdtime_respones",         required_argument, NULL, 'r' },
    { "sort_ascending",            no_argument,       NULL, 's' },
    { "sort_descending",           no_argument,       NULL, 'S' },
    { "verbose",                   no_argument,       NULL, 'v' },
    { "noverbose",                 no_argument,       NULL, 'V' },
    { "nosetup_commandline",       no_argument,       NULL, 1   },
    { "ns_commandline",            no_argument,       NULL, 1   },
    { NULL }
  };

  if (argc >= 1)
    strn0cpy(plugin_name, argv[0], sizeof(plugin_name));

  for (c = 1; c < argc; c++)
    dsyslog("%s: parameter%d=%s", plugin_name, c, argv[c]);

  #define Setvalue(T) T.c = true; T.u
  #define SetvalueChar(T) T.c = true; strn0cpy(T.u, optarg, sizeof(T.u))

  while ((c = getopt_long(argc, argv, "bBd:eEh:mMn:r:sSvV", long_options, NULL)) != -1)
  {
    // var pos4
    switch (c)
    {
      case 1:   commandline_preference.h = true;
                break;
/*      case 'b': Setvalue(delete_b_e) |= 0x1;
                break;
      case 'B': Setvalue(delete_b_e) &= ~0x1;
                break;
      case 'd': if (isnumber(optarg))
                {
                  Setvalue(num_entrys) = atoi(optarg);
                  Setvalue(delete_b_e) = 0x3;
                  if (num_entrys.u > 0) break;
                }
                esyslog("%s: invalid parameter for -d option: %s", plugin_name, optarg);
                return false;
      case 'e': Setvalue(delete_b_e) |= 0x2;
                break;
      case 'E': Setvalue(delete_b_e) &= ~0x2;
                break;
      case 'h': if (isnumber(optarg))
                {
                  Setvalue(message_in_queue) = atoi(optarg);
                  if (message_in_queue.u >=5 && message_in_queue.u <= 999) break;
                }
                esyslog("%s: invalid parameter for -h option: %s range 5-999", plugin_name, optarg);
                return false;
      case 'm': Setvalue(mainmenu_visible) = true;
                break;
      case 'M': Setvalue(mainmenu_visible) = false;
                break;
      case 'n': SetvalueChar(mainmenu_name);
                Setvalue(mainmenu_visible) = true;
                break;
      case 'r': if (isnumber(optarg))
                {
                  Setvalue(response_in_queue) = atoi(optarg);
                  if (response_in_queue.u >= 2 && response_in_queue.u <= 199) break;
                }
                esyslog("%s: invalid parameter for -r option: %s range 2-199", plugin_name, optarg);
                return false;
      case 's': Setvalue(sortdirection) = true;
                break;
      case 'S': Setvalue(sortdirection) = false;
                break;*/
      case 'v': Setvalue(verbose) = true;
                break;
      case 'V': Setvalue(verbose) = false;
                break;
      default:  return false;
    }
  }

  #undef Setvalue
  #undef SetvalueChar

  if (optind < argc && argv[optind][0] == '@')
  {
    strn0cpy(playlistconfigfile.u, &argv[optind][1], sizeof(playlistconfigfile.u));
    optind++;
  }
  return optind >= argc;
}

const char *cPluginPlaylist::Version(void)
{
  return VERSION;
}

const char *cPluginPlaylist::Description(void)
{
  return tr(DESCRIPTION);
}

const char *cPluginPlaylist::CommandLineHelp(void)
{
  // Return a string that describes all known command line options.
  return "  -m           --visible_in_mainmenu          Show the plugin in the mainmenu\n"
         "  -M           --hide_in_mainmenu             Hide the plugin in the mainmenu\n"
         "  -n Name      --mainmenu_name=Name           Select Name for entry in the mainmenu (set also -m)\n"
         "  -b           --delete_begin                 Show the delete all messages line at begin of messagelist\n"
         "  -B           --nodelete_begin               Hide the delete all messages line at begin of messagelist\n"
         "  -e           --delete_end                   Show the delete all messages line at end of messagelist\n"
         "  -E           --nodelete_end                 Hide the delete all messages line at end of messagelist\n"
         "  -d xx        --min_entrys=xx                Minimum entries for display delete all messages line at begin and end (include -b and -e)\n"
         "  -h xx        --holdtime_history=xx          Minimum time (min) for message in history queue (OSD-list) 5-999\n"
         "  -r xx        --holdtime_respones=xx         Minimum time (min) for responses in queue (readable by SAQRESP) 2-199\n"
         "  -s           --sort_ascending               Sort order for messagelist (OSD)\n"
         "  -S           --sort_descending              Sort order for messagelist (OSD)\n"
         "  -v           --verbose                      Enable more logging\n"
         "  -V           --noverbose                    Disable more loggig\n"
         "               --nosetup_commandline          Hide the Prefer Command Line Parameter from setup menu\n"
         "               --ns_commandline               Same as --nosetup_commandline\n";
         // free: a c f g i j k l o p q t u v w x y z
}

bool cPluginPlaylist::ProcessArgs(int argc, char *argv[])
{
  // Implement command line argument processing here if applicable.
  return ProcessArg(argc, argv);
}

#ifdef PL_Debug1
void TestI18n(char *Test)
{
  Setup.OSDLanguage = 0;
  const char *eng = tr(Test);
  Setup.OSDLanguage = 1;
  const char *deu = tr(Test);
  if (eng && deu && strcmp(eng, deu) == 0)
    dsyslog("%s: missing: %s", plugin_name, Test);
}
#endif

bool cPluginPlaylist::Start(void)
{
  // Start any background activities the plugin shall perform.
  TestAndSetOSDLanguage();

#ifdef PL_Debug1
  dsyslog("%s: cPluginPlaylist::Start playlistconfigfile=%s", plugin_name, playlistconfigfile.u);
#endif

  char *p;
  char *q = strdup(ConfigDirectory(""));
  asprintf(&p, "%s%s%s", q, *(q + strlen(q) - 1) == '/' ? "" : "/", playlistconfigfile.u);
  if (!access(playlistconfigfile.u, F_OK) && !access(playlistconfigfile.u, R_OK) || !access(p, F_OK) && !access(p, R_OK))
  {
    char *s;
    #define MAXARGS 100
    int fargc = 1;
    char *fargv[MAXARGS];
    cReadLine ReadLine;
    bool done;
    FILE *f;

    if (!access(playlistconfigfile.u, F_OK))
    {
      f = fopen(playlistconfigfile.u, "r");
#ifdef PL_Debug2
      dsyslog("%s: cPluginPlaylist::Start open playlistconfigfile=%s", plugin_name, playlistconfigfile.u);
#endif
    } else
    {
      f = fopen(p, "r");
#ifdef PL_Debug2
      dsyslog("%s: cPluginPlaylist::Start open playlistconfigfile=%s", plugin_name, p);
#endif
    }
    free(p);
    free(q);
    if (!f)
    {
      esyslog("%s: ERROR: cannot open config file: [%s]%s", plugin_name, ConfigDirectory(""), playlistconfigfile.u);
      return false;
    }
    while ((s = ReadLine.Read(f)) != NULL)
    {
      p = skipspace(stripspace(s));
      q = NULL;
      done = false;
      while (!done)
      {
        if (!q)
          q = p;
        switch (*p)
        {
          case '\\': strcpy(p, p + 1);
                     if (*p)
                       p++;
                     else
                     {
                       esyslog("%s: ERROR: missing character after \\", plugin_name);
                       return false;
                     }
                     break;
          case '"':
          case '\'': if ((p = SkipQuote(p)) == NULL)
                       return false;
                     break;
          default:   if (!*p || isspace(*p))
                     {
                       done = !*p;
                       *p = 0;
                       if (q)
                       {
                         if (fargc < MAXARGS - 1)
                         {
                           if (*q != '#')
                             fargv[fargc++] = strdup(q);
                         } else
                         {
                           esyslog("%s: ERROR: plugin argument list too long", plugin_name);
                           return false;
                         }
                         q = NULL;
                       }
                     }
                     if (!done)
                       p = *p ? p + 1 : skipspace(p + 1);
        }
      }
    }
    fclose(f);
    fargv[0] = strdup(plugin_name);
    fargv[fargc] = NULL;
    optind = 0; // to reset the getopt() data
    if (fargc > 1)
      if (!ProcessArg(fargc, fargv))
      {
        esyslog("%s: ERROR: cannot parse config file: [%s]%s", plugin_name, ConfigDirectory(""), playlistconfigfile.u);
        return false;
      }
      while(fargc) free(fargv[--fargc]);
  } else
  {
    free(p);
    free(q);
    if (strcmp(playlistconfigfile.u, playlistconfigfile.d))
    {
      esyslog("%s: ERROR: config file not found: [%s]%s", plugin_name, ConfigDirectory(""), playlistconfigfile.u);
      return false;
    } else if (verbose.u)
      isyslog("%s: INFO: config file not found: [%s]%s", plugin_name, ConfigDirectory(""), playlistconfigfile.u);
  }

  // default parameter                --> d_
  // read parameter from commandline  --> c_  (data in u_)
  // value parameter from config-file --> s_
  // read parameter from config-file  --> r_
  // paramater used                   --> u_

  #define SetParam(T) { if (!T.c || (!commandline_preference.u && T.r)) T.u = (!T.c && !T.r) ? T.d : T.s; }
  #define SetParamFile(T) { if (!T.c || (!commandline_preference.u && T.r)) strcpy(T.u, (!T.c && !T.r) ? T.d : T.s); }

  commandline_preference.u = commandline_preference.r ? commandline_preference.s : commandline_preference.d;

  // var pos5
  SetParamFile (playlistconfigfile);
  SetParam     (charentry);
  SetParam     (charfolder);
  SetParam     (charrecord);
  SetParamFile (lastplaylist);
  SetParamFile (mainmenu_name);
  SetParam     (mainmenu_visible);
  SetParam     (confirmdelplentry);
  SetParam     (okstartplay);
  SetParam     (deleterecords);
  SetParamFile (pathplaylists);
  SetParam     (startoption);
  SetParam     (storeplaylist);
  SetParam     (timeoutreadrecords);
  SetParam     (verbose);
  for (int i = Option_confirmdeleterecord; i < Option_max; i++)
    SetParam   (PL_options[i]);

  #undef SetParam
  #undef SetParamFile

  if (verbose.u)
    isyslog("%s: Start", plugin_name);
  DisplaySetings();
  strcpy(lastplaylist.o, lastplaylist.u);
  return true;
}

const char *cPluginPlaylist::MainMenuEntry(void)
{
  return mainmenu_visible.u ? mainmenu_name.u : NULL;
}

cOsdObject *cPluginPlaylist::MainMenuAction(void)
{
#ifdef PL_Debug1
  dsyslog("%s: cPluginPlaylist::MainMenuAction", plugin_name);
#endif
  TestAndSetOSDLanguage();
  return new cControlMenu;
}

cMenuSetupPage *cPluginPlaylist::SetupMenu(void)
{
  TestAndSetOSDLanguage();
  return new cMenuSetupPlayList;
}


bool cPluginPlaylist::SetupParse(const char *Name, const char *Value)
{
  #define SetParam(T) { T.r = true; T.s = strtol(Value, NULL, 0); }
  #define SetParamChar(T) { T.r = true; temp = strtol(Value, NULL, 0); T.s = temp > 255 ? temp - 256 : temp; }
  #define SetParamFile(T) { T.r = true; strn0cpy(T.s, Value, sizeof(T.s)); }

  // Parse your own setup parameters and store their values.
  dsyslog("%s: Setupparameter %s=%s", plugin_name, Name, Value);
  // var pos6
  int temp;

  if      (!strcasecmp(Name, "cahre"))       SetParamChar (charentry)
  else if (!strcasecmp(Name, "charf"))       SetParamChar (charfolder)
  else if (!strcasecmp(Name, "charr"))       SetParamChar (charrecord)
  else if (!strcasecmp(Name, "commandline")) SetParam     (commandline_preference)
  else if (!strcasecmp(Name, "lastpl"))      SetParamFile (lastplaylist)
  else if (!strcasecmp(Name, "name"))        SetParamFile (mainmenu_name)
  else if (!strcasecmp(Name, "visible"))     SetParam     (mainmenu_visible)
  else if (!strcasecmp(Name, "deletentry"))  SetParam     (confirmdelplentry)
  else if (!strcasecmp(Name, "okplay"))      SetParam     (okstartplay)
  else if (!strcasecmp(Name, "delete"))      SetParam     (deleterecords)
  else if (!strcasecmp(Name, "path"))        SetParamFile (pathplaylists)
  else if (!strcasecmp(Name, "start"))       SetParam     (startoption)
  else if (!strcasecmp(Name, "storepl"))     SetParam     (storeplaylist)
  else if (!strcasecmp(Name, "timerecords")) SetParam     (timeoutreadrecords)
  else if (!strcasecmp(Name, "verbose"))     SetParam     (verbose)
  else if (!strcasecmp(Name, "confdelete"))  SetParam     (PL_options[Option_confirmdeleterecord])
  else if (!strcasecmp(Name, "startnew"))    SetParam     (PL_options[Option_confirmstartnewrecord])
  else if (!strcasecmp(Name, "deletel"))     SetParam     (PL_options[Option_deletefromlist])
  else if (!strcasecmp(Name, "deleter"))     SetParam     (PL_options[Option_deleterecord])
  else if (!strcasecmp(Name, "jumpmark"))    SetParam     (PL_options[Option_jumpmark])
  else if (!strcasecmp(Name, "playoption1")) SetParam     (PL_options[Option_playoption1])
  else if (!strcasecmp(Name, "playoption2")) SetParam     (PL_options[Option_playoption2])
  else if (!strcasecmp(Name, "playonlynew")) SetParam     (PL_options[Option_playonlynew])
  else if (!strcasecmp(Name, "searchnew"))   SetParam     (PL_options[Option_searchnewrecord])
  else if (!strcasecmp(Name, "searchsub"))   SetParam     (PL_options[Option_searchrecordsub])
  else
    return false;
  return true;

  #undef SetParam
  #undef SetParamChar
  #undef SetParamFile
}

const char *cPluginPlaylist::ExpandPath(const char *Filename, bool CreateDir)
{
  static char path[MaxFileName];
  char *c = strdup(ConfigDirectory(""));
  char *p = NULL;

  ExpandEnvironment(&pathplaylists);
  if (pathplaylists.e[0] == '/')
    asprintf(&p, "%s%s%s", pathplaylists.e, *(pathplaylists.e + strlen(pathplaylists.e) - 1) == '/' ? "" : "/", Filename);
  else
    asprintf(&p, "%s%s%s%s%s", c, *(c + strlen(c) - 1) == '/' ? "" : "/", pathplaylists.e, strlen(pathplaylists.e) ? *(pathplaylists.e + strlen(pathplaylists.e) - 1) == '/' ? "" : "/" : "", Filename);
  if (strlen(p) >= MaxFileName)
    esyslog("%s: max length for filename is %d this name is %d [%s]", plugin_name, MaxFileName, strlen(p), p);
  strn0cpy(path, p, MaxFileName);
  if (CreateDir)
    MakeDirs(path, false);
  free(c);
  free(p);
#ifdef PL_Debug1
  dsyslog("%s: ExpandPath Filename=%s Return=%s", plugin_name, Filename, path);
#endif
  return &path[0];
}

VDRPLUGINCREATOR(cPluginPlaylist); // Don't touch this!


vdrtools.c
/*
 * playlist: A plugin for the Video Disk Recorder
 *
 * See the README file for copyright information and how to reach the author.
 *
 * $Id: vdrtools.c 0.2 2004/10/08 02:34:00 hflor Exp $
 */

#include "vdrtools.h"
#include "playlist.h"
#include "dataplaylist.h"
#include <vdr/menu.h>
#include <vdr/interface.h>

char *SkipQuote(char *s)
{
  char c = *s;
  strcpy(s, s + 1);
  while (*s && *s != c)
  {
    if (*s == '\\')
      strcpy(s, s + 1);
    if (*s)
      s++;
  }
  if (*s)
  {
    strcpy(s, s + 1);
    return s;
  }
  esyslog("ERROR: missing closing %c", c);
  return NULL;
}

char *ExchangeChars(char *s, bool ToFileSystem)
{
// ToFileSystem is not use, only for comatibility with 'ExchangeChars' from 'recording.c'
  char *p = s;
  while (*p)
  {
    switch (*p)
    {
         // mapped characters:
      case '_':    *p = ' ';
                   break;
      case '/':    *p = '~';
                   break;
         // encodes characters:
      case '#':    if (strlen(p) > 2)
                   {
                     char buf[3];
                     sprintf(buf, "%c%c", *(p + 1), *(p + 2));
                     unsigned char c = strtol(buf, NULL, 16);
                     *p = c;
                     memmove(p + 1, p + 3, strlen(p) - 2);
                   }
                   break;
      // backwards compatibility:
      case '\x01': *p = '\'';
                   break;
      case '\x02': *p = '/';
                   break;
      case '\x03': *p = ':';
                   break;
    }
    p++;
  }
  return s;
}

bool DeleteRecording(const char *Filename, bool Confirm)
{
  if (!Confirm || Interface->Confirm(tr("Delete recording?")))
  {
    cRecordControl *rc = cRecordControls::GetRecordControl(Filename);
    if (rc)
    {
      if (!Confirm || Interface->Confirm(tr("Timer still recording - really delete?")))
      {
        cTimer *timer = rc->Timer();
        if (timer)
        {
          timer->Skip();
          cRecordControls::Process(time(NULL));
          if (timer->IsSingleEvent())
          {
            isyslog("timer %d deleted", timer->Index() + 1);
            Timers.Del(timer);
          }
          Timers.Save();
        }
      } else
        return false;
    }
    cRecording *recording = Recordings.GetByName(Filename);
    if (recording)
    {
      if (recording->Delete())
      {
        cReplayControl::ClearLastReplayed(Filename);
        if (LastSelectedRecord && !strcmp(LastSelectedRecord, Filename))
          FREENULL(LastSelectedRecord);
        Recordings.Del(recording);
        for (cPlaylist *pl = PlaylistCol.First(); pl; pl = PlaylistCol.Next(pl))
          for (cPlaylistRecord *plr = pl->First_PLR(); plr; plr = pl->Next_PLR(plr))
            if (strcasecmp(plr->Filename(), Filename) == 0)
              plr->SetDel();
        return true;
      }
    }
    ERROR(tr("Error while accessing recording!"));
  }
  return false;
}


#ifdef PL_Debug1
const char *KeyName(eKeys Key)
{
  switch (RAWKEY(Key))
  {
    case kUp:         return tr("Up");             break;
    case kDown:       return tr("Down");           break;
    case kMenu:       return tr("Menu");           break;
    case kOk:         return tr("Ok");             break;
    case kBack:       return tr("Back");           break;
    case kLeft:       return tr("Left");           break;
    case kRight:      return tr("Right");          break;
    case kRed:        return tr("Red");            break;
    case kGreen:      return tr("Green");          break;
    case kYellow:     return tr("Yellow");         break;
    case kBlue:       return tr("Blue");           break;
    case k0:          return "0";                  break;
    case k1:          return "1";                  break;
    case k2:          return "2";                  break;
    case k3:          return "3";                  break;
    case k4:          return "4";                  break;
    case k5:          return "5";                  break;
    case k6:          return "6";                  break;
    case k7:          return "7";                  break;
    case k8:          return "8";                  break;
    case k9:          return "9";                  break;
    case kPlay:       return tr("Play");           break;
    case kPause:      return tr("Pause");          break;
    case kStop:       return tr("Stop");           break;
    case kRecord:     return tr("Record");         break;
    case kFastFwd:    return tr("FastFwd");        break;
    case kFastRew:    return tr("FastRew");        break;
    case kPower:      return tr("Power");          break;
    case kChanUp:     return tr("ChanUp");         break;
    case kChanDn:     return tr("ChanDn");         break;
    case kVolUp:      return tr("VolUp");          break;
    case kVolDn:      return tr("VolDn");          break;
    case kMute:       return tr("Mute");           break;
    case kSchedule:   return tr("Schedule");       break;
    case kChannels:   return tr("Channels");       break;
    case kTimers:     return tr("Timers");         break;
    case kRecordings: return tr("Recordings");     break;
    case kSetup:      return tr("Setup");          break;
    case kCommands:   return tr("Commands");       break;
    case kUser1:      return tr("User1");          break;
    case kUser2:      return tr("User2");          break;
    case kUser3:      return tr("User3");          break;
    case kUser4:      return tr("User4");          break;
    case kUser5:      return tr("User5");          break;
    case kUser6:      return tr("User6");          break;
    case kUser7:      return tr("User7");          break;
    case kUser8:      return tr("User8");          break;
    case kUser9:      return tr("User9");          break;
    case kNone:       return tr("none");           break;
    case kKbd:        return tr("Kbd");            break;
    default:          return tr("unknow");         break;
  }
}

const char *OSStateName(eOSState OSState)
{
  switch (OSState)
  {
    case osUnknown:    return "osUnknown";
    case osContinue:   return "osContinue";
    case osSchedule:   return "osSchedule";
    case osChannels:   return "osChannels";
    case osTimers:     return "osTimers";
    case osRecordings: return "osRecordings";
    case osPlugin:     return "osPlugin";
    case osSetup:      return "osSetup";
    case osCommands:   return "osCommands";
    case osPause:      return "osPause";
    case osRecord:     return "osRecord";
    case osReplay:     return "osReplay";
    case osStopRecord: return "osStopRecord";
    case osStopReplay: return "osStopReplay";
    case osCancelEdit: return "osCancelEdit";
    case osSwitchDvb:  return "osSwitchDvb";
    case osBack:       return "osBack";
    case osEnd:        return "osEnd";
    case os_User:      return "os_User";
    case osUser1:      return "osUser1";
    case osUser2:      return "osUser2";
    case osUser3:      return "osUser3";
    case osUser4:      return "osUser4";
    case osUser5:      return "osUser5";
    case osUser6:      return "osUser6";
    case osUser7:      return "osUser7";
    case osUser8:      return "osUser8";
    case osUser9:      return "osUser9";
    case osUser10:     return "osUser10";
#if VDRVERSNUM < 10307
    case osUser11:     return "osUser11";
#endif
    default:           return "unknow";
  }
  return "";
}
#endif
